source/numerics/matrix_tools.cc

```diff
@@ -2559,12 +2557,6 @@ namespace MatrixTools
             Assert (local_range == solution.local_range(),
                     ExcInternalError());
 
-            // we have to read and write from this
-            // matrix (in this order). this will only
-            // work if we compress the matrix first,
-            // done here
-            matrix.compress ();
-
```

Comment:
With this change and the change in clear_rows, code that was valid before can now throw an exception.